generate_network.py

```diff
@@ -105,79 +95,110 @@ def generate_scalefree_graph(n):
+# Label the generated network with prefix
+def postprocess_network(G):
+    G = nx.Graph(G)                             # prune out parallel/multi edges
+    G.remove_edges_from(nx.selfloop_edges(G))   # remove the self-loops
     mapping = {i: f"{prefix}{i}" for i in range(len(G))}
-    return nx.relabel_nodes(G, mapping)
+    return nx.relabel_nodes(G, mapping)         # label the nodes
+
 
+### file format related fns ###########################################################
+
+#Generate per node toml configs
+def generate_toml(topics, node_type=node_types.desktop):
+    topic_str =  " ". join(get_random_sublist(topics))    # topics as a space separated string
+    if node_type == node_type.desktop:
+        toml = "rpc-admin = true\nkeep-alive = true\n"
+    elif node_type == node_type.mobile:
+        toml =  "rpc-admin = true\nkeep-alive = true\n"
+    else:
+        print(node_type +": Unsupported node type")
+        sys.exit(1)
+    toml += f"topics = \"{topic_str}\"\n"
+    return toml
 
-# Generate dump data from the network and topics
-def generate_dump_data(H, topics):
-    data_to_dump = {}
-    global ports_shifted
+
+# Generates network-wide json and per-node toml and writes them 
+def generate_and_write_files(dirname, num_topics, H):
+    if not os.path.exists(dirname):
+        os.mkdir(dirname)
+    elif not os.path.isfile(dirname) and os.listdir(dirname):
+        print(dirname +": exists and is not empty")
+        sys.exit(1)
+    elif os.path.isfile(dirname):
+        print(dirname +": exists and is not a directory")
+        sys.exit(1)
+
+    topics = generate_topics(num_topics)
+    json_dump = {}
     for node in H.nodes:
-        data_to_dump[node] = {}
-        data_to_dump[node]["ports-shift"] = ports_shifted
-        ports_shifted += 1
-        data_to_dump[node]["topics"] = get_random_sublist(topics)
-        data_to_dump[node]["static-nodes"] = []
+        write_toml(dirname, node, generate_toml(topics))        # per node toml
+        json_dump[node] = {}
+        json_dump[node]["static-nodes"] = []
         for edge in H.edges(node):
-            data_to_dump[node]["static-nodes"].append(edge[1])
-    return data_to_dump
+            json_dump[node]["static-nodes"].append(edge[1])
+    write_json(dirname, json_dump)                              # network wide json
 
 
-def main():
-    #extract the CLI arguments and assign params
-    options = get_options()
-    fname = options.fname
-    num_nodes = options.num_nodes
-    num_topics = options.num_topics
-    nw_type = options.nw_type
-    prefix = "waku_"     
-    num_partitions = options.num_partitions
-    #num_edges = options.num_edges    ## need to control num_edges?
+### the main ###########################################################
+def main(
+        dirname: str = "Waku", num_nodes: int = 3, num_topics: int = 1, 
+        nw_type: nw_types = "NEWMANWATTSSTROGATZ", 
+        node_type: node_types = "DESKTOP",
```

Comment:
Not sure, but maybe you can use the enum value directly instead of the strings:

```python
...
nw_type: NwType = NwType.NewmanWhatever
...
```

Reply:
I did try it as the first thing, but typer complained rather loudly. typer would put OCaml to shame on incomprehensible error messages! Will try doing it again.

Reply:
No worries then, I think it need to implement something to be used like that, don't remember what. But that should be ok anyway :)

Reply:
looks like typer expects EnumType.EnumeeEntry.value for assignments; and want them to be strings. So did nw_type : networkType = networkType.NewmanWhatever.value